test/lit/passes/inlining_optimize-level=3.wast

```diff
@@ -220,5 +230,197 @@
+  ;; CHECK-NEXT:   (local.set $1
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (block (result i32)
+  ;; CHECK-NEXT:    (call $recursive-inlining-1
+  ;; CHECK-NEXT:     (local.get $1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $recursive-inlining-1 (export "A") (param $x i32) (result i32)
+    (call $recursive-inlining-2
+      (local.get $x)
+    )
+  )
+
+  ;; CHECK:      (func $recursive-inlining-2 (param $x i32) (result i32)
+  ;; CHECK-NEXT:  (call $recursive-inlining-1
+  ;; CHECK-NEXT:   (local.get $x)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $recursive-inlining-2 (export "B") (param $x i32) (result i32)
+    (call $recursive-inlining-1
+      (local.get $x)
+    )
+  )
+
+  ;; As above, but both call the second. The first will inline the second
+  ;; several times before hitting the limit (of around 5).
+
+  ;; CHECK:      (func $b-recursive-inlining-1 (param $x i32) (result i32)
+  ;; CHECK-NEXT:  (local $1 i32)
+  ;; CHECK-NEXT:  (local $2 i32)
+  ;; CHECK-NEXT:  (local $3 i32)
+  ;; CHECK-NEXT:  (local $4 i32)
+  ;; CHECK-NEXT:  (local $5 i32)
+  ;; CHECK-NEXT:  (block $__inlined_func$b-recursive-inlining-2 (result i32)
+  ;; CHECK-NEXT:   (local.set $1
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (block (result i32)
+  ;; CHECK-NEXT:    (block $__inlined_func$b-recursive-inlining-20 (result i32)
+  ;; CHECK-NEXT:     (local.set $2
+  ;; CHECK-NEXT:      (local.get $1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (block (result i32)
+  ;; CHECK-NEXT:      (block $__inlined_func$b-recursive-inlining-21 (result i32)
+  ;; CHECK-NEXT:       (local.set $3
+  ;; CHECK-NEXT:        (local.get $2)
+  ;; CHECK-NEXT:       )
+  ;; CHECK-NEXT:       (block (result i32)
+  ;; CHECK-NEXT:        (block $__inlined_func$b-recursive-inlining-22 (result i32)
+  ;; CHECK-NEXT:         (local.set $4
+  ;; CHECK-NEXT:          (local.get $3)
+  ;; CHECK-NEXT:         )
+  ;; CHECK-NEXT:         (block (result i32)
+  ;; CHECK-NEXT:          (block $__inlined_func$b-recursive-inlining-23 (result i32)
+  ;; CHECK-NEXT:           (local.set $5
+  ;; CHECK-NEXT:            (local.get $4)
+  ;; CHECK-NEXT:           )
+  ;; CHECK-NEXT:           (call $b-recursive-inlining-2
+  ;; CHECK-NEXT:            (local.get $5)
+  ;; CHECK-NEXT:           )
+  ;; CHECK-NEXT:          )
+  ;; CHECK-NEXT:         )
+  ;; CHECK-NEXT:        )
+  ;; CHECK-NEXT:       )
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $b-recursive-inlining-1 (export "BA") (param $x i32) (result i32)
+    (call $b-recursive-inlining-2
+      (local.get $x)
+    )
+  )
+
+  ;; CHECK:      (func $b-recursive-inlining-2 (param $x i32) (result i32)
+  ;; CHECK-NEXT:  (call $b-recursive-inlining-2
+  ;; CHECK-NEXT:   (local.get $x)
+  ;; CHECK-NEXT:  )
```

Comment:
Which part of the pass prevents this call from being inlined into itself?

Reply:
(same as above, the rule against self-inlining)